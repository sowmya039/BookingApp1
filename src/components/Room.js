import React, { Component } from 'react';
import {
  StyledRoom,
  StyledRoomInner,
  StyledInput,
  StyledRoomName,
  StyledSelect
} from './styles';

export default class Room extends Component {
  _onChange = (ev) => {
    this.props._onChange(ev, this.props.index);
  };
  render() {
    const { room = {}, index } = this.props;
    return (
      <StyledRoom index>
        {index > 0 ? (
          <StyledInput
            type="checkbox"
            checked={room.selected}
            onChange={this._onChange}
          />
        ) : null}
        <StyledRoomName checked={room.selected}>{room.name}</StyledRoomName>

        <StyledRoomInner checked={room.selected}>
          <div>
            <p>Adults</p>
            <p>(18+)</p>
            <StyledSelect
              className="styled-select"
              value={room.adults}
              disabled={!room.selected}
              onChange={this._onChange}
              name="adults"
            >
              {[1, 2].map((item, index) => {return (
                <option value={item} key={`${index}`}>
                  {item}
                </option>
              );})}
            </StyledSelect>
          </div>
          <div>
            <p>Children</p>
            <p>(0-17)</p>
            <StyledSelect
              className="styled-select"
              value={room.children}
              disabled={!room.selected}
              onChange={this._onChange}
              name="children"
            >
              {[0, 1, 2].map((item, index) => {return (
                <option value={item} key={`${index}`}>
                  {item}
                </option>
              );})}
            </StyledSelect>
          </div>
        </StyledRoomInner>
      </StyledRoom>
    );
  }
}
