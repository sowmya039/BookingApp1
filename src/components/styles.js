import styled from 'styled-components';
/*Main container section*/
export const StyledContainer = styled.section`
  padding: 10px;
`;

export const StyledApp = styled.div`
  font-family: sans-serif;
  font-size: 100%;
`;
/* styles for each room section*/
export const StyledRoom = styled.section`
  width: 10%;
  min-width: 12.5rem;
  margin: 1.2rem;
  border-radius: 0.625rem;
  border: 2px solid #ccc;
  background-color: #ddd;
  display: inline-block;
  text-align: ${(props) => {return (props.index > 0 ? 'center' : '');}};
`;
/* section for Adults and children divs*/
export const StyledRoomInner = styled.section`
  padding: 15px;
  display: flex;
  justify-content: space-around;
  border-radius: 0px 0px 10px 10px;
  background-color: ${(props) => {return (props.checked ? 'white' : '');}};
`;
/*checkbox*/
export const StyledInput = styled.input`
  float: left;
  margin-top: 1.25rem;
  margin-left: 3.125rem;
  width: 1.25rem;
  height: 1.25rem;
  border: 2px solid #ccc;
`;

/* when room is not selected/unchecked, room name shoud not be bold */
export const StyledRoomName = styled.h4`
  font-weight: ${(props) => {return (props.checked ? '' : 'normal');}};
`;

/* To increase the size of the select option*/
export const StyledSelect = styled.select`
  height: 25px;
  width: 40px;
  font-size: 15px;
`;
/*Input submit button*/
export const Input = styled.input`
  background-color: #b8bdb8;
  color: rgb(10, 10, 10);
  padding: 10px;
  margin-left: 12px;
  width: 6rem;
  border-radius: 3px;
  cursor: pointer;
  font-weight: bold;
  display: inline-block;
  border: none;

  @media only screen and (max-width: 1200px) {
    width: 15%;
  }
  @media only screen and (max-width: 1023px) {
    width: 18%;
  }

  @media only screen and (max-width: 700px) {
    width: 25%;
  }
  @media only screen and (max-width: 320px) {
    width: 25%;
  }
  @media only screen and (max-width: 290px) {
    width: 35%;
  }
  @media only screen and (max-width: 290px) {
    width: 40%;
  }
  @media only screen and (max-width: 290px) {
    width: 45%;
  }
`;
export const StyledClear = styled.div`
  clear: both;
`;
