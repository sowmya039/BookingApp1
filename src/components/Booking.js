import React, { Component } from 'react';
import Room from './Room';
import { connect } from 'react-redux';
import { getSavedRooms, updateRooms, saveRooms } from '../actions';
import { StyledContainer, StyledClear, Input } from './styles';

export class Booking extends Component {
  constructor() {
    super();
    this._onChange = this._onChange.bind(this);
    this._onSubmit = this._onSubmit.bind(this);
  }
  componentDidMount() {
    this.props.getSavedRooms();
  }
  _onChange(ev, index) {
    this.props.updateRooms(ev, index);
  };

  _onSubmit(ev) {
    ev.preventDefault();
    this.props.saveRooms();
  };
  render() {
    const { rooms = [] } = this.props;
    return (
      <StyledContainer data-test="styledcontainer">
        <form onSubmit={this._onSubmit} data-test="form-element">
          {rooms.map((room, index) => {
            return (
              <Room
                data-test="room-element"
                room={room}
                key={`${index}`}
                index={index}
                _onChange={this._onChange}
              />
            );
          })}
          <StyledClear />
          <Input type="submit" value="Submit" />
        </form>
      </StyledContainer>
    );
  }
}

const mapStateToProps = (state) => {
  return {
    rooms: state
  };
};
const mapDispatchToProps = (dispatch) => {
  return {
    getSavedRooms: () => {return dispatch(getSavedRooms());},
    updateRooms: (ev, index) => {return dispatch(updateRooms(ev, index));},
    saveRooms: () => {return dispatch(saveRooms());}
  };
};
const connectedBooking = connect(
  mapStateToProps,
  mapDispatchToProps
)(Booking);
export default connectedBooking;
