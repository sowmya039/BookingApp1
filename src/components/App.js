import React from 'react';
import { Provider } from 'react-redux';
import configureStore from '../store';
import Booking from './Booking';
import './styles.js';
export default () => {
  const store = configureStore();
  return (
    <Provider store={store}>
      <Booking />
    </Provider>
  );
};
