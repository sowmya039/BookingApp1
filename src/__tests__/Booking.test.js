import React from 'react';
import { shallow, mount } from 'enzyme';
import { Booking } from '../components/Booking';
import sinon from 'sinon';
import toJson from 'enzyme-to-json';
const [getSavedRooms, updateRooms, saveRooms] = new Array(3).fill(jest.fn());

function setup(isFull = false) {
    const props = {
        rooms: [
            {
                name: 'Room 1',
                selected: true,
                adults: 1,
                children: 0
            },
            {
                name: 'Room 2',
                selected: false,
                adults: 1,
                children: 0
            },
            {
                name: 'Room 3',
                selected: false,
                adults: 1,
                children: 0
            },
            {
                name: 'Room 4',
                selected: false,
                adults: 1,
                children: 0
            }
        ],
        getSavedRooms,
        updateRooms,
        saveRooms
    };
    const renderFunc = isFull ? mount : shallow;
    const enzymeWrapper = renderFunc(<Booking {...props} />);
    return {
        props,
        enzymeWrapper
    };
}

describe('Shallow/full-mount of Booking App', () => {
    it('StyledContainer,form element should be available', () => {
        const {enzymeWrapper } = setup();
        expect(enzymeWrapper.find('[data-test="styledcontainer"]').length).toBe(1);
        expect(enzymeWrapper.find('[data-test="form-element"]').length).toBe(1);
    });
    it('should contain n rooms as specified in props', () => {
        const { props, enzymeWrapper } = setup(true);
        expect(enzymeWrapper.find('[data-test="room-element"]').length).toBe(props.rooms.length);
    });
    it('room checkbox should be checked/unchecked based on props', () => {
        const { props, enzymeWrapper } = setup(true);
        const roomElms = enzymeWrapper.find('[data-test="room-element"]');
        expect(roomElms.at(1).find('input').find('[type="checkbox"]').prop('checked')).toBe(props.rooms[1].selected);
    });
    it('room name should match with props value', () => {
        const { props, enzymeWrapper } = setup(true);
        const roomElms = enzymeWrapper.find('[data-test="room-element"]');
        expect(roomElms.at(0).find('h4').text()).toBe(props.rooms[0].name);
    });
});

describe('method calls', () => {
    let wrapper;
    beforeEach(() => {
        sinon.spy(Booking.prototype, '_onChange');
        sinon.spy(Booking.prototype, '_onSubmit');
        const { enzymeWrapper } = setup(true);
        wrapper = enzymeWrapper;
    });
    afterEach(() => {
        Booking.prototype._onSubmit.restore();
        Booking.prototype._onChange.restore();
    });
    it('should invoke _onChange', () => {
        const roomElms = wrapper.find('[data-test="room-element"]');
        const checkBox = roomElms.at(1).find('input').find('[type="checkbox"]');
        checkBox.simulate('change');
        expect(Booking.prototype._onChange.calledOnce).toBe(true);
    });
});

describe('Snapshot Testing', () => {
    it('should match snapshot when shallow', () => {
        const { enzymeWrapper } = setup();
        expect(toJson(enzymeWrapper)).toMatchSnapshot();
    });
});