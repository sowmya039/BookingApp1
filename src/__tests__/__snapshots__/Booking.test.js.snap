// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Snapshot Testing should match snapshot when shallow 1`] = `
<styled.section
  data-test="styledcontainer"
>
  <form
    data-test="form-element"
    onSubmit={[Function]}
  >
    <Room
      _onChange={[Function]}
      data-test="room-element"
      index={0}
      key="0"
      room={
        Object {
          "adults": 1,
          "children": 0,
          "name": "Room 1",
          "selected": true,
        }
      }
    />
    <Room
      _onChange={[Function]}
      data-test="room-element"
      index={1}
      key="1"
      room={
        Object {
          "adults": 1,
          "children": 0,
          "name": "Room 2",
          "selected": false,
        }
      }
    />
    <Room
      _onChange={[Function]}
      data-test="room-element"
      index={2}
      key="2"
      room={
        Object {
          "adults": 1,
          "children": 0,
          "name": "Room 3",
          "selected": false,
        }
      }
    />
    <Room
      _onChange={[Function]}
      data-test="room-element"
      index={3}
      key="3"
      room={
        Object {
          "adults": 1,
          "children": 0,
          "name": "Room 4",
          "selected": false,
        }
      }
    />
    <styled.div />
    <styled.input
      type="submit"
      value="Submit"
    />
  </form>
</styled.section>
`;
